core/drivers/dcp/imx_dcp.c

```diff
@@ -0,0 +1,1035 @@
+		desc->ctrl1 |= DCP_CONTROL1_KEY_SELECT_OTP_CRYPTO;
+	} else if (init->key_mode == PAYLOAD) {
+		desc->ctrl0 |= DCP_CONTROL0_PAYLOAD_KEY;
+		aes->payload_size += DCP_AES128_KEY_SIZE;
+		memcpy(aes->buffer, init->key, DCP_AES128_KEY_SIZE);
+	} else {
+		desc->ctrl1 |= SHIFT_U32(init->key_mode, 8);
+	}
+
+	if (init->mode == CBC) {
+		desc->ctrl0 |= DCP_CONTROL0_CIPHER_INIT | DCP_CONTROL0_CHAIN;
+		desc->ctrl1 |= DCP_CONTROL1_CIPHER_MODE_CBC;
+		memcpy(aes->buffer + aes->payload_size, init->iv,
+		       DCP_AES128_IV_SIZE);
+		aes->payload_size += DCP_AES128_IV_SIZE;
+	}
+
+	desc->src_buffer = NULL;
+	desc->dest_buffer = NULL;
+	desc->status = 0;
+	desc->buff_size = 0;
+	desc->next = virt_to_phys(desc);
+	aes->op_state.init = true;
+
+	return ret;
+}
+
+TEE_Result dcp_aes_do_update(struct aesdata *aes, const uint8_t *src,
+			     uint8_t *dst, size_t size)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	struct align_buf output = {};
+	unsigned int nb_block = 0;
+	unsigned int idx = 0;
+	uint8_t data_block[DCP_AES_BLOCK_SIZE] = {};
+	struct dcp_descriptor *desc = NULL;
+
+	if (!aes) {
+		EMSG("Bad parameters, aes is NULL");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+	/* Check if dcp_aes_do_init() has been called before */
+	if (!aes->op_state.init) {
+		EMSG("Error, please call dcp_aes_do_init() before");
+		return TEE_ERROR_ACCESS_DENIED;
+	}
+	if (size % DCP_AES128_BUFFER_SIZE || !size) {
+		EMSG("Bad parameters, size has to be a multiple of 16bytes");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+	if (!src) {
+		EMSG("Bad parameters, src is NULL");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+	if (!dst) {
+		EMSG("Bad parameters, dst is NULL");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	ret = dcp_calloc_align_buf(&output, size);
+	if (ret != TEE_SUCCESS) {
+		EMSG("Output aligned buffer allocation failed");
+		return ret;
+	}
+
+	desc = &aes->dcp_data.desc;
+
+	nb_block = size / DCP_AES_BLOCK_SIZE;
+	if (!nb_block) {
+		nb_block = 1;
+		desc->buff_size = size;
+	} else {
+		desc->buff_size = DCP_AES_BLOCK_SIZE;
+	}
+
+	/* Lock the channel */
+	ret = dcp_lock_channel(&aes->op_state, &aes->dcp_data.channel);
+	if (ret != TEE_SUCCESS) {
+		EMSG("Channel is busy, can't start operation now");
+		goto exit;
+	}
+
+	desc->src_buffer = data_block;
+
+	for (idx = 0; idx < nb_block; idx++) {
+		memcpy(data_block, src + idx * DCP_AES_BLOCK_SIZE,
+		       DCP_AES_BLOCK_SIZE);
+
+		desc->dest_buffer = output.data + idx * DCP_AES_BLOCK_SIZE;
+		desc->payload = aes->buffer;
+
+		if (aes->payload_size)
+			cache_operation(TEE_CACHECLEAN,
+					aes->dcp_data.desc.payload,
+					aes->payload_size);
+
+		ret = dcp_run(&aes->dcp_data);
+
+		if (ret != TEE_SUCCESS)
+			break;
+
+		if (desc->ctrl0 & DCP_CONTROL0_CIPHER_INIT) {
+			desc->ctrl0 &= ~DCP_CONTROL0_CIPHER_INIT;
+			aes->payload_size -= DCP_AES128_IV_SIZE;
+		}
+	}
+
+	if (ret == TEE_SUCCESS)
+		memcpy(dst, output.data, output.size);
+
+	/* Unlock the channel */
+	dcp_unlock_channel(aes->dcp_data.channel);
+
+exit:
```

Comment:
We'd normally call this label out `out:`

Reply:
I renamed it to ```out```